docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -56,42 +89,90 @@ The current multisig public keys use amino serialization to generate the address
 those public keys and their address formatting, and call them "legacy amino" multisig public keys
 in protobuf. We will also create multisig public keys without amino addresses to be described below.
 
+### Hash Function Choice
+
+We propose to use [blake2b](https://www.blake2.net/) as a hash function choice:
++ The main arguments are speed and separating from `sha256` which is widely used
+  by miners and could potentially be used to find collisions.
++ The function was in the final round of the 2012 NIST hash function competition.
++ It's well studied with security covered in many academic papers.
++ Faster than `sha2` on non ASICs chipsets.
++ It's getting more traction in other blockchains (Pokadot, Sia, Zcash, ...). Related [zcash discussion](https://github.com/zcash/zcash/issues/706#issuecomment-187807410).
++ It's already widely supported by all major programming languages.
++ Cryptography consulting reviled no argument against `blake2b`
+
+
+### Base Address Algorithm
+
+We start with defining a base algorithm for generating addresses. Notably, it's used for Base Accounts (accounts represented by a single key-pair) addresses. For each Public Key schema we need to have an associated `typ` string, which we will discuss in a section below. `hash` is a cryptographic hash function defined in the previous section.
+
+```go
+const A_LEN = 32
+
+func BaseAddress(typ string, pubkey []byte) []byte {
+    return hash(hash(typ) + pubkey)[:A_LEN]
+}
+```
+
+The `+` is a bytes concatenation, which doesn't use any separator.
+
+This algorithm is an outcome after a consulting session with a cryptographer.
+Motivation: this algorithm keeps the address relatively small (length of the `typ` doesn't impact on the length of the final address)
+and it's more secure than [post-hash-prefix-proposal] (with reducing the pubkey hash to 20 bytes, we essentially significantly reduce the address space).
+Moreover the cryptographer motivated the choice to add `typ` in the hash to protect against switch table attack.
 
-### Canonical Address Format
 
-We have three types of accounts we would like to create addresses for in the future:
-- regular public key addresses for new signature algorithms (ex. `sr25519`).
-- public key addresses for multisig public keys that don't use amino encoding
-- module accounts: basically any accounts which cannot sign transactions and
-which are managed internally by modules
```

Comment:
why is this deleted?

Reply:
I reworded it and moved it to the section above.